### PR TITLE
Counters are always generated content and cannot be selected

### PR DESCRIPTION
**Pull request by Ahmad-S792**
#### dbcaeff80040bbfadeee5162d44a7e8ee7415b81
<pre>
Counters are always generated content and cannot be selected

Counters are always generated content and cannot be selected
<a href="https://bugs.webkit.org/show_bug.cgi?id=249500">https://bugs.webkit.org/show_bug.cgi?id=249500</a>

Reviewed by Ryosuke Niwa.

This patch is to align Webkit behavior with Blink / Chromium and Gecko / Firefox.

Merge - <a href="https://chromium.googlesource.com/chromium/src.git/+/9d8a90f8a7f49c673bcd4e369efc84c62cc18285">https://chromium.googlesource.com/chromium/src.git/+/9d8a90f8a7f49c673bcd4e369efc84c62cc18285</a>

This patch is modify the overriding behavior where RenderCounter inherits the logic of CanBeSelectionLeaf(), which returns always true, from RenderText. This is to modify to explicitly return false for
RenderCount as it is always generated content.

* Source/WebCore/rendering/RenderCount.h: Add bool &quot;CanBeSelectionLeaf&quot; with return value as &quot;false&quot;
* LayoutTests/editing/selection/text-selection-counter.html: Add Test Case
* LayoutTests/editing/selection/text-selection-counter-expected.html: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: Updated to skip added test on iOS platform due to selection support

Canonical link: <a href="https://commits.webkit.org/258080@main">https://commits.webkit.org/258080@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/407a93f56c795d5cfda70dcb65f14de311fbe4f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100847 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9997 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33894 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110150 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170424 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104836 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10929 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/853 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93258 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107995 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106630 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8259 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91520 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34874 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22916 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77851 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3687 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24435 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3706 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/813 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9817 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43935 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5540 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5484 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by rniwa** [comment]
I don't understand why we want to make this code change. What's the end user benefit?

**Comment by Ahmad-S792** [comment]
> I don't understand why we want to make this code change. What's the end user benefit?

Based on the test case, I linked in bug, it improves the user experience and it does not let counters be selected, which is behavior matching Chrome and Firefox. [Test Case Link](https://jsfiddle.net/u7r2qw0f/show) - Try to select the text and it will show selection on counter to select the text, while same test does not show selection for counter in Chrome and Firefox.

**Comment by Ahmad-S792** [comment]
@rniwa - I checked via JSFiddle and this test even fail on my personal iPhone device, it seems that repaint does not work on iOS properly and I have checked TestExceptions and there are quite a few repaint tests skipped on iOS. Should I add platform specific exception or you have any idea or suggestion?

File: LayoutTests/editing/selection/text-selection-counter.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
+<script>
+function selectRangeAfterLayoutAndPaint(startElement, startIndex, endElement, endIndex) {
```
Comment (rniwa) [inline LayoutTests/editing/selection/text-selection-counter.html]: Looks like this function is never used?

**Comment by rniwa** [comment]
On iOS, selection painting goes through a completely different code paths so I suspect this test wouldn't work there. I suggest we skip the test on iOS.

File: LayoutTests/editing/selection/text-selection-counter.html (hunk 1 above)
Comment (Ahmad-S792) [reply LayoutTests/editing/selection/text-selection-counter.html]: Yes - after removing, the test was working fine, so I removed it.

**Comment by Ahmad-S792** [comment]
> On iOS, selection painting goes through a completely different code paths so I suspect this test wouldn't work there. I suggest we skip the test on iOS.

I added it to Skip next to other editing/selection test case. :-)

**Review by rniwa — APPROVED** [review]

**Pull request merged**